cEP-0020.md

```diff
@@ -0,0 +1,355 @@
+| cEP      | 20                                                     |
+| Version  | 1.0                                                    |
+| Title    | Newcomer metrics and Gamification system               |
+| Authors  | Shrikrishna Singh <mailto:krishnasingh.ss30@gmail.com> |
+| Status   | Proposed                                               |
+| Type     | Process                                                |
+
+## Abstract
+
+This cEP describes how a gamification system will automate the process
+of becoming a `Developer` from a `Newcomer`, as a part of this
+[GSoC project](https://summerofcode.withgoogle.com/projects/#5892040252981248).
+
+For the implementation of this process, we will incorporate a gamification
+system which will allow us to track newcomer's progress and assign some
+points to each of the activity they do. We will also have different levels
+and give them badges on the basis of points earned and performed activities
+by the newcomers.
+
+So, every newcomer at coala will have to complete the processes involved in
+the gamification system in order to become a developer at coala or to get the
+**coala Developer** badge.
+
+## Background
+
+### Newcomer Experience
+
+**The problems we have:**
+
+- How to attract more newcomers
+  - How to cope with the increasing amounts of newcomers
+- How to teach and help newcomers to learn faster so
+  we can keep working on coala at a high pace.
+- How to improve newcomer -> dev/maintainer ratio
+- How to increase amount of feedback we get
+  - How to use feedback to improve coala and processes
+
+## Why Gamification?
+
+As coala tries to be a welcoming organization for newcomers, it gives them
+clear pathways and as much direct assistance as possible. However, it is
+already known that newcomers face many barriers while attempting to contribute
+to the open source community for the first time. Some barriers they face include
+orientation issues that can potentially demotivate newcomers from placing
+their first contribution. On the other hand, gamification is widely used to
+engage and motivate people to accomplish tasks and improve their performance.
+Therefore, the goal of this project is to use gamification to orient and
+motivate newcomers to overcome onboarding barriers to contribute and engage
+with coala and its community.
+
+## Gamification Elements
+
+There will be three game design elements to help newcomers: Points, Levels
+and Badges. For each of these elements, we will have a set of rules to
+describe the operation of the gamified environment. i.e.:
+
+### Points and Rules
+
+Let’s start by defining some rules for earning points. Points are a
+simple tool to reward behavior, and they provide an excellent base upon which to
+build a gamified design. The first step is to define the activities that we want
+newcomers to take. i.e.:
+
+- Created Issues by a Newcomer
+- Created Pull Requests by a Newcomer
```

Comment:
Misusing by making excessive PRs (knowingly or unknowingly)?

Reply:
Ya we need to track closed-not-merged PRs

Reply:
@prnvdixit @sks444 , yes we need to think about abuse.  I've mentioned it frequently, but it isnt dealt with in detail in the cEP yet, probably as the bottom half hasnt been overhauled yet.

The problem is where do we record abusing behaviour so that this system can use it.  Obtaining a feed of feedback from corobo is possible approach, but corobo is currently not storing state, and I doubt we would do that.

Meta-review provides a lot of opportunity for storing state, but it doesnt cover everything.  It especially does not cover gitter/IM behaviour.
We may want to look at replies to a newcomer IM messages as a way to create a feedback loop there.

One concern with the schema design below is that the points dont seem to link to specific actions, so it wont be possible to use the data to see which actions counted and which ones didnt, which makes it harder to diagnose why a score isnt as high as it should be, which becomes especially relevant if there are rules aimed at preventing scores for potentially abuse behaviour.

Reply:
> One concern with the schema design below is that the points dont seem to link to specific actions, so it wont be possible to use the data to see which actions counted and which ones didnt, which makes it harder to diagnose why a score isnt as high as it should be, 

This gets me to thinking - Shouldn't there be any logs being maintained for all increments/decrements of points (I'm thinking of a possibility if something goes wrong, we can just use (maybe manually) some deferred or immediate modification recovery)?